Fix Crash When Running Multiple Reductions in the Preview Tab

cailafinn opened the pull request:
**Description of work.**
Add some notifying calls between the job runner and the preview view to disable the tab while the reduction is running, which was causing some race-based segfaults if the load button was spammed. 

The cause of the segfaults appeared to be when two reductions were started very close together. When the second one finished, the other tried to access a validator that the first one had already cleared, so was trying to deref null pointers. 

**To test:**
1. Open the ISIS Reflectometry interface
2. Go to the Preview tab
3. Enter run `INTER45455_inst` and click Load (This data is available on this Epic and must be in your user search path: #31069)
4. Put an angle of 0 and try to click load a bunch. The button should be disabled while processing. The algo should error, but not crash.
5. Now put an angle in and click load a lot again. The line plot should be filled in and it shouldn't crash. 
6. Now draw a square on the inst viewer and move your drawn area as many times as you can in the shortest time to try and force reductions to overlap. The tab should disable itself while the reduction is running and prohibit you from doing this. 

Fixes #34865 

*This does not require release notes* because **this tab isn't user facing yet.**

<!-- Ensure the base of this PR is correct (e.g. release-next or main)
Finally, don't forget to add the appropriate labels, milestones, etc.!  -->

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Are the release notes saved in a separate file, using Issue or PR number for file name and in the correct location?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

Commit:
Disable load widgets while loading

Currently won't renable if the algorithm fails. That comes next.

This is to fix a crash that occurred if enter or the load button was
pressed to quickly after pressing it the first time.

RE #34865

Commit:
Add method to enable widgets on load failure

Still need to add the call to this from the model.

RE #34865

Commit:
Add call to model

RE #34865

Commit:
Enable widgets only after reduction complete

Allowing loading after the workspace has been loaded isn't effective as
asking to run a load while the *reduction* was running was what caused
the crash. Instead we enable the load widgets after:
- A successful reduction
- A failed workspace load
- A failed reduction

RE #34865

sf1919 commented:
Have set the linux & systems test to run separately - https://builds.mantidproject.org/job/pull_requests-conda-linux/2281/

**For any reviewer/gatekeeper:**
The link in the checks list won't pick up on this re-set test so it will stay as pending. Provided the job in the link above passes please review as normal.

Commit:
Replace load widget enable control with whole tab

Basically anything that re-runs the reduction has the potential to cause
a crash, it's just a bit harder to do than just spamming on yhe load
button. Insead of worrying about all of the different buttons that can
cause problems individually, let's just disable all the widgets on the
tab until the reduction is complete. That should essentially erase the
danger of segfaults caused by racing algorithms that have been started
as it seems to be the clearing of validators that was causing seg-faults
all along.

RE #34745

Commit:
Remove unused method

RE #34865

rbauststfc commented on qt/scientific_interfaces/ISISReflectometry/GUI/Preview/PreviewPresenter.cpp:
```diff
@@ -74,16 +74,17 @@ void PreviewPresenter::notifyAutoreductionPaused() { updateWidgetEnabledState();
 void PreviewPresenter::notifyLoadWorkspaceRequested() {
+  m_view->disableMainWidget();
   auto const name = m_view->getWorkspaceName();
   try {
```
If I try to load `POLREF37125` (which is not a matrix workspace and therefore throws an error) the tab stays disabled and I can't attempt to load a different run. I'm not sure if it's anything to do with not re-enabling the view as part of this try/catch, or if the issue is elsewhere.

Similarly, when testing with `INTER45455_inst`, if `ReflectometryISISSumBanks` throws an error then the tab doesn't seem to be re-enabled.

rbauststfc reviewed (changes requested):
This does a great job of preventing the crashes and the code changes look good to me 👍

I just noticed a couple of things around re-enabling the tab when there are errors and keeping the Update button disabled when the angle value hasn't changed.

rbauststfc committed:
Merge branch 'main' into 34865_repeat_load_crash

rbauststfc committed:
Ensure Update button is only enabled when angle is changed

RE #34865

rbauststfc committed:
Ensure tab re-enabled on runtime and SumBanks errors

RE #34865

rbauststfc committed:
Notify algorithm error when update function throws

RE #34865

rbauststfc reviewed (dismissed):
I've resolved the conflicts and added a few changes to address the things I raised in my first review. I'm going to approve this, but @gemmaguest it would probably be good for you to gatekeeper this one to check the changes I've made.

gemmaguest reviewed (dismissed):
I've tested this quite thoroughly and on the whole it is looking really good. 

One problem though - it looks like it has introduced a bug where ROIs can no longer be deleted! Whenever I click to select an ROI (in either the instrument viewer or sliceviewer) the reduction gets re-run, even if I didn't actually change the ROI. I suspect that because of that it is somehow "forgetting" that the ROI is selected ready for editing?

ConorMFinn commented:
> One problem though - it looks like it has introduced a bug where ROIs can no longer be deleted! Whenever I click to select an ROI (in either the instrument viewer or sliceviewer) the reduction gets re-run, even if I didn't actually change the ROI. I suspect that because of that it is somehow "forgetting" that the ROI is selected ready for editing?

Yep, that's the reason. I actually found this before I went off for Xmas and started looking into fixing it but forgot to re-draft the PR! Sorry!

Commit:
Add getter for selected region to model

RE #34865

Commit:
Only run reduction if region changed

RE #34865

Commit:
Deactivate selectors after running reduction

RE #34865

Commit:
Only re sum banks when they are changed

This essentially means don't rerun the reduction and the bank summing
algorithm unless the position of one of the selectors is changed or
deleted.

RE #34865

rbauststfc reviewed (commented):
The recent code changes have fixed the issue with deleting ROIs. 👍

I'm not sure if I'm testing incorrectly, but it looks to me like a reduction is now only run when the first ROI added to the slice viewer is created and/or moved? Adding extra background or transmission ROIs doesn't look to me like it's triggering a re-run of the reduction any more?

Commit:
Fix errors when deleting regions

RE #34865

ConorMFinn commented:
> I'm not sure if I'm testing incorrectly, but it looks to me like a reduction is now only run when the first ROI added to the slice viewer is created and/or moved? Adding extra background or transmission ROIs doesn't look to me like it's triggering a re-run of the reduction any more?

Good catch on this. Turns out there were some logic errors in what I'd written that helped to reveal a few other problems I hadn't noticed before (such as the model starting off with `std::optional`s without a value on initialisation but having a value but empty vectors when all the regions of that type are deleted. I've now added the additional logic to make that comparison work properly.

rbauststfc commented on qt/scientific_interfaces/ISISReflectometry/GUI/Preview/PreviewPresenter.cpp:
```diff
@@ -204,7 +217,25 @@ void PreviewPresenter::notifyRegionChanged() {
+         isRegionChanged(m_model->getSelectedRegion(ROIType::Background),
+                         m_regionSelector->getRegion(roiTypeToString(ROIType::Background))) ||
+         isRegionChanged(m_model->getSelectedRegion(ROIType::Transmission),
+                         m_regionSelector->getRegion(roiTypeToString(ROIType::Transmission)));
```
Just a suggestion on style in case you agree - I was wondering if the below alternative might help cut down some of the duplication in `PreviewPresenter::isRegionSelectionChanged`, although at the expense of a bit more boilerplate (I haven't checked the below, so apologies if it doesn't work!):

```
bool PreviewPresenter::isRegionSelectionChanged() {
  return isRegionChanged(ROIType::Signal) ||
         isRegionChanged(ROIType::Background) ||
         isRegionChanged(ROIType::Transmission);
}

bool PreviewPresenter::isRegionChanged(ROIType roiType) {
  auto oldValue = m_model->getSelectedRegion(roiType);
  auto newValue = m_regionSelector->getRegion(roiTypeToString(roiType));
  return !(!oldValue.has_value() && newValue.empty()) && oldValue != newValue;
}
```

Also, I'm trying to decide, is `!(!modelValue.has_value() && viewValue.empty())` equivalent to `(modelValue.has_value() || !viewValue.empty())` - i.e. either the old or new ROI has to have a value?

rbauststfc reviewed (commented):
The new logic looks like it's working really well. I see what you mean about the tricky boolean logic, great job figuring that out! 👍

I'm happy to approve as is, but I had a couple of thoughts on style that I thought I'd run by you first, just in case you agree.

Commit:
Refactor to remove boilerplace

RE #34865

Co-authored-by: Rachel Baust  <rbauststfc@users.noreply.github.com>

rbauststfc reviewed (approved):
The tab seems to be working well, the crashes have all been sorted and the code changes look good 👍👍

rbauststfc commented:
@gemmaguest I think this one would be good for you to gatekeeper if possible, as the "change requested" status is left over from your last review.
